File: plank.py
import cv2
import mediapipe as mp
import math
import time
from collections import deque

UPPER_THRESHOLD = 165
LOWER_THRESHOLD = 155
SMOOTH_WINDOW = 10
DEBOUNCE_FRAMES = 5

mp_pose = mp.solutions.pose
mp_hands = mp.solutions.hands
mp_draw = mp.solutions.drawing_utils

angle_buffer = deque(maxlen=SMOOTH_WINDOW)
plank_buffer = deque(maxlen=DEBOUNCE_FRAMES)
waiting_for_start = True
plank_started = False
start_time = 0.0
hold_time = 0.0

# ─── Gesture Detection ──────────────────────────────────────────────────

def detect_start_movement(results_pose, results_hands):
    if not results_pose.pose_landmarks or not results_hands.multi_hand_landmarks:
        return False
    nose_y = results_pose.pose_landmarks.landmark[0].y
    for hand in results_hands.multi_hand_landmarks:
        wrist_y = hand.landmark[mp_hands.HandLandmark.WRIST].y
        if abs(wrist_y - nose_y) < 0.1:
            return True
    return False

def detect_reset_movement(results_hands):
    if not results_hands.multi_hand_landmarks or len(results_hands.multi_hand_landmarks) != 2:
        return False
    wrist1 = results_hands.multi_hand_landmarks[0].landmark[mp_hands.HandLandmark.WRIST]
    wrist2 = results_hands.multi_hand_landmarks[1].landmark[mp_hands.HandLandmark.WRIST]
    dx = abs(wrist1.x - wrist2.x)
    return dx > 0.3

# ─── Angle Helper ────────────────────────────────────────────────────────

def calculate_angle(a, b, c):
    ba = (a[0] - b[0], a[1] - b[1])
    bc = (c[0] - b[0], c[1] - b[1])
    cosang = (ba[0]*bc[0] + ba[1]*bc[1]) / (math.hypot(*ba) * math.hypot(*bc))
    cosang = max(min(cosang, 1.0), -1.0)
    return math.degrees(math.acos(cosang))

# ─── Main Loop ───────────────────────────────────────────────────────────

cap = cv2.VideoCapture(0)
with mp_pose.Pose(min_detection_confidence=0.5, min_tracking_confidence=0.5) as pose, \
     mp_hands.Hands(max_num_hands=2, min_detection_confidence=0.5, min_tracking_confidence=0.5) as hands:

    while True:
        ret, frame = cap.read()
        if not ret:
            break

        image = cv2.cvtColor(cv2.flip(frame, 1), cv2.COLOR_BGR2RGB)
        results_pose = pose.process(image)
        results_hands = hands.process(image)
        image = cv2.cvtColor(image, cv2.COLOR_RGB2BGR)

        # ── Gesture Logic ──
        if detect_reset_movement(results_hands):
            waiting_for_start = True
            plank_started = False
            hold_time = 0.0
            angle_buffer.clear()
            plank_buffer.clear()
            print("🔁 Reset detected")

        elif waiting_for_start and detect_start_movement(results_pose, results_hands):
            waiting_for_start = False
            start_time = time.time()
            plank_started = False
            hold_time = 0.0
            print("🟢 Start detected")

        # ── Visual Feedback ──
        if waiting_for_start:
            cv2.putText(image, "✋ Hand on head to start plank", (10, 50),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.8, (255, 255, 255), 2)
        else:
            raw_angle = None
            if results_pose.pose_landmarks:
                lm = results_pose.pose_landmarks.landmark
                def hip_angle(side):
                    s = side.upper()
                    sh = [lm[getattr(mp_pose.PoseLandmark, f"{s}_SHOULDER")].x,
                          lm[getattr(mp_pose.PoseLandmark, f"{s}_SHOULDER")].y]
                    hi = [lm[getattr(mp_pose.PoseLandmark, f"{s}_HIP")].x,
                          lm[getattr(mp_pose.PoseLandmark, f"{s}_HIP")].y]
                    kn = [lm[getattr(mp_pose.PoseLandmark, f"{s}_KNEE")].x,
                          lm[getattr(mp_pose.PoseLandmark, f"{s}_KNEE")].y]
                    return calculate_angle(sh, hi, kn)

                left_ang = hip_angle("LEFT")
                right_ang = hip_angle("RIGHT")
                raw_angle = (left_ang + right_ang) / 2.0
                cv2.putText(image, f"{int(raw_angle)}°", (10, 50),
                            cv2.FONT_HERSHEY_SIMPLEX, 1, (255, 255, 255), 2)

            if raw_angle is not None:
                angle_buffer.append(raw_angle)
                avg_angle = sum(angle_buffer) / len(angle_buffer)
                if not plank_started:
                    plank_buffer.append(avg_angle > UPPER_THRESHOLD)
                else:
                    plank_buffer.append(avg_angle > LOWER_THRESHOLD)
            else:
                plank_buffer.append(False)

            is_plank = len(plank_buffer) == DEBOUNCE_FRAMES and all(plank_buffer)

            if is_plank and not plank_started:
                plank_started = True
            elif not is_plank and plank_started:
                plank_started = False

            if plank_started:
                hold_time = time.time() - start_time

            if results_pose.pose_landmarks:
                mp_draw.draw_landmarks(image, results_pose.pose_landmarks, mp_pose.POSE_CONNECTIONS)

            mins, secs = divmod(int(hold_time), 60)
            timer_text = f"{mins:02d}:{secs:02d}"
            cv2.rectangle(image, (0, 0), (220, 90), (0, 0, 0), -1)
            cv2.putText(image, "PLANK", (10, 30),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.8, (0, 255, 255), 2)
            cv2.putText(image, timer_text, (10, 80),
                        cv2.FONT_HERSHEY_SIMPLEX, 2, (255, 255, 255), 2)

        cv2.imshow("Plank Timer", image)
        if cv2.waitKey(10) & 0xFF == ord('q'):
            break

cap.release()
cv2.destroyAllWindows()

File: plank.js
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Plank Timer</title>
  <script src="https://cdn.jsdelivr.net/npm/@mediapipe/pose/pose.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/@mediapipe/hands/hands.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/@mediapipe/camera_utils/camera_utils.js"></script>
</head>
<body>
  <video id="video" playsinline style="display:none;"></video>
  <canvas id="canvas"></canvas>
  <script>
    const videoElement = document.getElementById('video');
    const canvasElement = document.getElementById('canvas');
    const ctx = canvasElement.getContext('2d');

    // Config
    const UPPER_THRESHOLD = 165;
    const LOWER_THRESHOLD = 155;
    const SMOOTH_WINDOW   = 15;
    const DEBOUNCE_FRAMES = 10;

    // State buffers
    const angleBuffer = [];
    const plankBuffer = [];
    let waitingForStart = true;
    let plankStarted = false;
    let startTime = 0;
    let holdTime = 0;
    let lastPose = null;
    let lastHands = null;

    // Initialize Pose
    const pose = new Pose.Pose({
      locateFile: (file) => `https://cdn.jsdelivr.net/npm/@mediapipe/pose/${file}`,
      modelComplexity: 1,
      enableSegmentation: false,
      smoothLandmarks: true,
      minDetectionConfidence: 0.5,
      minTrackingConfidence: 0.5
    });
    pose.onResults(onResults);

    // Initialize Hands
    const hands = new Hands.Hands({
      locateFile: (file) => `https://cdn.jsdelivr.net/npm/@mediapipe/hands/${file}`,
      maxNumHands: 2,
      minDetectionConfidence: 0.5,
      minTrackingConfidence: 0.5
    });
    hands.onResults(onResults);

    // Camera setup
    const camera = new Camera.Camera(videoElement, {
      onFrame: async () => {
        await hands.send({ image: videoElement });
        await pose.send({ image: videoElement });
      },
      width: 640,
      height: 480
    });
    camera.start();

    function onResults(results) {
      if (results.poseLandmarks) lastPose = results;
      if (results.multiHandLandmarks) lastHands = results;
      if (!lastPose) return;

      // Draw the frame
      canvasElement.width = videoElement.videoWidth;
      canvasElement.height = videoElement.videoHeight;
      ctx.save();
      ctx.drawImage(videoElement, 0, 0);
      // Draw pose skeleton
      if (lastPose.poseLandmarks) {
        drawConnectors(ctx, lastPose.poseLandmarks, Pose.POSE_CONNECTIONS);
        drawLandmarks(ctx, lastPose.poseLandmarks);
      }

      // Detect reset gesture (hands far apart)
      if (lastHands.multiHandLandmarks && lastHands.multiHandLandmarks.length >= 2) {
        const w1 = lastHands.multiHandLandmarks[0][HandLandmark.WRIST];
        const w2 = lastHands.multiHandLandmarks[1][HandLandmark.WRIST];
        if (Math.abs(w1.x - w2.x) > 0.30) {
          resetState();
        }
      }

      // Detect start gesture (hand near nose)
      if (waitingForStart && lastHands.multiHandLandmarks) {
        const nose = lastPose.poseLandmarks[Pose.PoseLandmark.NOSE];
        for (const hand of lastHands.multiHandLandmarks) {
          const wrist = hand[HandLandmark.WRIST];
          if (wrist.y < nose.y + 0.02) {
            waitingForStart = false;
            startTime = performance.now();
            angleBuffer.length = 0;
            plankBuffer.length = 0;
          }
        }
      }

      // Display prompt or timer
      if (waitingForStart) {
        drawText('Perform start gesture', 10, 50, 'yellow');
      } else {
        trackPlank();
      }

      ctx.restore();
    }

    function trackPlank() {
      // Compute hip angle
      const lm = lastPose.poseLandmarks;
      const avgAngle = ((angle(lm, 'LEFT') + angle(lm, 'RIGHT')) / 2);
      drawText(`Ang: ${avgAngle.toFixed(0)}°`, 10, 100, 'white');

      // Smooth + debounce
      angleBuffer.push(avgAngle);
      if (angleBuffer.length > SMOOTH_WINDOW) angleBuffer.shift();
      const smoothAngle = angleBuffer.reduce((a,b) => a+b)/angleBuffer.length;
      if (!plankStarted) plankBuffer.push(smoothAngle > UPPER_THRESHOLD);
      else plankBuffer.push(smoothAngle > LOWER_THRESHOLD);
      if (plankBuffer.length > DEBOUNCE_FRAMES) plankBuffer.shift();

      // Toggle after consistent reads
      if (!plankStarted && plankBuffer.every(v => v)) {
        plankStarted = true;
        startTime = performance.now();
        plankBuffer.length = 0;
      } else if (plankStarted && plankBuffer.every(v => !v)) {
        plankStarted = false;
        plankBuffer.length = 0;
      }

      // Update hold time
      if (plankStarted) {
        holdTime = (performance.now() - startTime) / 1000;
      }

      // Draw timer UI
      const mins = String(Math.floor(holdTime/60)).padStart(2, '0');
      const secs = String(Math.floor(holdTime % 60)).padStart(2, '0');
      ctx.fillStyle = 'black'; ctx.fillRect(0,0,250,90);
      drawText('PLANK', 10, 30, 'lime');
      drawText(`${mins}:${secs}`, 10, 80, 'white');
    }

    function angle(lm, side) {
      const S = Pose.PoseLandmark;
      const sh = lm[S[side + '_SHOULDER']];
      const hi = lm[S[side + '_HIP']];
      const kn = lm[S[side + '_KNEE']];
      return calculate_angle([sh.x, sh.y], [hi.x, hi.y], [kn.x, kn.y]);
    }

    function resetState() {
      waitingForStart = true;
      plankStarted = false;
      holdTime = 0;
      angleBuffer.length = 0;
      plankBuffer.length = 0;
    }

    function drawText(text, x, y, color) {
      ctx.fillStyle = color;
      ctx.font = '20px sans-serif';
      ctx.fillText(text, x, y);
    }
  </script>
</body>
</html>

